operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1728 @@
+	cell.In
+
+	Logger logrus.FieldLogger
+
+	LC hive.Lifecycle
+
+	ClientSet    k8sClient.Clientset
+	PoolResource resource.Resource[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+	SvcResource  resource.Resource[*core_v1.Service]
+
+	OperatorConfig *operatorOption.OperatorConfig
+	Config         LBIPAMConfig
+
+	NotifyReadyCallbacks []LBIPAMInitDone `group:"lb-ipam-init-ready"`
+}
+
+func newLBIPAM(params LBIPAMParams) *LBIPAM {
+	if !params.Config.EnableLBIPAM {
+		return nil
+	}
+
+	var lbClasses []string
+	if params.OperatorConfig.BGPAnnounceLBIP {
+		lbClasses = append(lbClasses, "io.cilium/bgp-control-plane")
+	}
+
+	lbIPAM := &LBIPAM{
+		logger:            params.Logger,
+		poolResource:      params.PoolResource,
+		svcResource:       params.SvcResource,
+		initDoneCallbacks: params.NotifyReadyCallbacks,
+		poolClient:        params.ClientSet.CiliumV2alpha1().CiliumLoadBalancerIPPools(),
+		svcClient:         params.ClientSet.CoreV1(),
+		poolStore:         NewPoolStore(),
+		rangesStore:       newRangesStore(),
+		serviceStore:      NewServiceStore(),
+		lbClasses:         lbClasses,
+		ipv4Enabled:       option.Config.IPv4Enabled(),
+		ipv6Enabled:       option.Config.IPv6Enabled(),
+	}
+
+	// Make a new Ctx which we will use to manage the lifetime of LBIPAM.Run.
+	ipamCtx, cancel := context.WithCancel(context.Background())
+	var wg sync.WaitGroup
+
+	params.LC.Append(hive.Hook{
+		OnStart: func(ctx hive.HookContext) error {
+			params.Logger.Info("Starting LB IPAM")
+
+			wg.Add(1)
+
+			// Start go routine, we can't block in OnStart
+			go func() {
+				defer wg.Done()
+				lbIPAM.Run(ipamCtx)
+			}()
+
+			return nil
+		},
+		OnStop: func(ctx hive.HookContext) error {
+			// Close the ipamCtx
+			cancel()
+
+			// Convert blocking Wait into a channel so we can select on the context closing or the Wait.
+			wgDone := make(chan struct{})
+			go func() {
+				wg.Wait()
+				close(wgDone)
+			}()
+
+			select {
+			case <-wgDone:
+			case <-ctx.Done():
+			}
+
+			return nil
+		},
+	})
+
+	return lbIPAM
+}
+
+// LBIPAM is the loadbalancer IP address manager, watcher/controller which allocates and assigns IP addresses
+// to LoadBalancer services from the configured set of LoadBalancerIPPools in the cluster.
+type LBIPAM struct {
+	logger logrus.FieldLogger
+
+	lbClasses   []string
+	ipv4Enabled bool
+	ipv6Enabled bool
+
+	poolClient cilium_client_v2alpha1.CiliumLoadBalancerIPPoolInterface
+	svcClient  client_typed_v1.ServicesGetter
+
+	poolResource resource.Resource[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+	svcResource  resource.Resource[*core_v1.Service]
+
+	poolStore    poolStore
+	rangesStore  rangesStore
+	serviceStore serviceStore
+
+	ctx context.Context
+
+	// Only used during testing.
+	initDoneCallbacks []LBIPAMInitDone
+}
+
+type ipPoolEvent = resource.Event[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+type svcEvent = resource.Event[*core_v1.Service]
+
+func (ipam *LBIPAM) Run(ctx context.Context) {
+	ipam.ctx = ctx
```

Comment:
Why not assign it in `newLBIPAM`? I'd prefer that, or not have `ctx` in `LBIPAM` at all and just pass it down from `Run()`